File: app/src/main/java/com/example/fruitsdiary/usecase/addeditentry/AddEditEntryAdapter.java
package com.example.fruitsdiary.usecase.addeditentry;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.fruitsdiary.R;
import com.example.fruitsdiary.databinding.ViewFruitEntryBinding;
import com.example.fruitsdiary.model.FruitEntry;
import com.example.fruitsdiary.util.StringUtils;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class AddEditEntryAdapter extends RecyclerView.Adapter<AddEditEntryAdapter.FruitEntryViewHolder> {

    private List<FruitEntry> mFruitEntryList;
    private OnItemClickListener mOnItemClickListener;

    AddEditEntryAdapter(OnItemClickListener onItemClickListener) {
        mFruitEntryList = new ArrayList<>();
        mOnItemClickListener = onItemClickListener;
    }

    void setFruitEntryList(@NonNull List<FruitEntry> fruitEntryList) {
        mFruitEntryList = fruitEntryList;
        notifyDataSetChanged();
    }

    void removeFruitEntry(FruitEntry fruitEntry) {
        int index = getFruitEntryIndex(fruitEntry);
        mFruitEntryList.remove(index);
        notifyItemRemoved(index);
    }

    private int getFruitEntryIndex(FruitEntry fruitEntry) {
        return mFruitEntryList.indexOf(fruitEntry);
    }

    @NonNull
    @Override
    public FruitEntryViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        ViewFruitEntryBinding binding = DataBindingUtil.inflate(
                LayoutInflater.from(viewGroup.getContext()),
                R.layout.view_fruit_entry,
                viewGroup,
                false
        );
        return new FruitEntryViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull FruitEntryViewHolder fruitEntryViewHolder, int i) {
        FruitEntry fruitEntry = mFruitEntryList.get(i);
        ViewFruitEntryBinding binding = fruitEntryViewHolder.binding;
        Context context = binding.getRoot().getContext();

        setFruitImage(context, fruitEntry.getImage(), binding.fruitImage);
        setFruitName(context, fruitEntry, binding.fruit);
        setVitamins(context, fruitEntry, binding.vitamins);

        fruitEntryViewHolder.bind(fruitEntry);

        if (fruitEntry.isModified()) {
            binding.getRoot().setBackgroundColor(ContextCompat.getColor(context, R.color.light_grey));
        } else {
            binding.getRoot().setBackgroundColor(Color.WHITE);
        }

    }

    private void setFruitImage(Context context, String fruitImagePath, ImageView fruitImageView) {
        String fruitImageUrl = String.format(
                context.getString(R.string.fruit_image_url),
                fruitImagePath
        );

        Picasso.get()
                .load(fruitImageUrl)
                .placeholder(R.drawable.fruit_image_placeholder)
                .into(fruitImageView);
    }

    private void setFruitName(Context context, FruitEntry fruitEntry, TextView fruitTextView) {
        String fruitName = StringUtils.getCorrectFruitSpelling(
                context,
                fruitEntry.getAmount(),
                fruitEntry.getType()
        );
        fruitTextView.setText(
                String.format(StringUtils.getFruitNumberFormat(), fruitEntry.getAmount(), fruitName)
        );
    }

    private void setVitamins(Context context, FruitEntry fruitEntry, TextView vitaminTextView) {
        int vitamins = fruitEntry.getVitamins() * fruitEntry.getAmount();
        String vitaminText = context.getResources().getQuantityString(
                R.plurals.vitamins,
                vitamins
        );
        vitaminTextView.setText(
                String.format(StringUtils.getVitaminNumberFormat(), vitamins, vitaminText)
        );
    }

    @Override
    public int getItemCount() {
        return mFruitEntryList.size();
    }

    class FruitEntryViewHolder extends RecyclerView.ViewHolder {

        private ViewFruitEntryBinding binding;

        FruitEntryViewHolder(@NonNull ViewFruitEntryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(@NonNull final FruitEntry fruitEntry) {
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    mOnItemClickListener.onItemClick(fruitEntry);
                }
            });
        }
    }

    public interface OnItemClickListener {
        void onItemClick(@NonNull FruitEntry fruitEntry);
    }
}

File: app/src/main/java/com/example/fruitsdiary/usecase/addeditentry/AddEditEntryAdapter.kt
package com.example.fruitsdiary.usecase.addeditentry

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.fruitsdiary.R
import com.example.fruitsdiary.databinding.ViewFruitEntryBinding
import com.example.fruitsdiary.model.FruitEntry
import com.example.fruitsdiary.util.StringUtils
import com.squareup.picasso.Picasso
import java.util.*

class AddEditEntryAdapter (private val mOnItemClickListener: OnItemClickListener) : RecyclerView.Adapter<AddEditEntryAdapter.FruitEntryViewHolder>() {

    var fruitEntryList: MutableList<FruitEntry> = ArrayList()
        set(fruitList) {
            field = fruitList
            notifyDataSetChanged()
        }

    fun removeFruitEntry(fruitEntry: FruitEntry) {
        val index = getFruitEntryIndex(fruitEntry)
        fruitEntryList.removeAt(index)
        notifyItemRemoved(index)
    }

    private fun getFruitEntryIndex(fruitEntry: FruitEntry): Int {
        return fruitEntryList.indexOf(fruitEntry)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): FruitEntryViewHolder {
        val binding = DataBindingUtil.inflate<ViewFruitEntryBinding>(
                LayoutInflater.from(viewGroup.context),
                R.layout.view_fruit_entry,
                viewGroup,
                false
        )
        return FruitEntryViewHolder(binding)
    }

    override fun onBindViewHolder(fruitEntryViewHolder: FruitEntryViewHolder, i: Int) {
        val fruitEntry = fruitEntryList[i]
        val binding = fruitEntryViewHolder.binding
        val context = binding.root.context

        setFruitImage( fruitEntry.image, binding.fruitImage)
        setFruitName(context, fruitEntry, binding.fruit)
        setVitamins(context, fruitEntry, binding.vitamins)

        fruitEntryViewHolder.bindClick(fruitEntry)

        if (fruitEntry.isModified) {
            binding.root.setBackgroundColor(ContextCompat.getColor(context, R.color.light_grey))
        } else {
            binding.root.setBackgroundColor(Color.WHITE)
        }

    }

    private fun setFruitImage(fruitImagePath: String?, fruitImageView: ImageView) {
        Picasso.get()
                .load(fruitImagePath)
                .placeholder(R.drawable.fruit_image_placeholder)
                .into(fruitImageView)
    }

    private fun setFruitName(context: Context, fruitEntry: FruitEntry, fruitTextView: TextView) {
        val fruitName = StringUtils.getCorrectFruitSpelling(
                context,
                fruitEntry.amount,
                fruitEntry.type
        )
        fruitTextView.text = "${fruitEntry.amount}  $fruitName"
    }

    private fun setVitamins(context: Context, fruitEntry: FruitEntry, vitaminTextView: TextView) {
        val vitamins = fruitEntry.vitamins * fruitEntry.amount
        val vitaminText = context.resources.getQuantityString(
                R.plurals.vitamins,
                vitamins
        )
        vitaminTextView.text = "$vitamins $vitaminText"
    }

    override fun getItemCount(): Int {
        return fruitEntryList.size
    }

    inner class FruitEntryViewHolder(val binding: ViewFruitEntryBinding) : RecyclerView.ViewHolder(binding.root) {

        fun bindClick(fruitEntry: FruitEntry) {
            itemView.setOnClickListener { mOnItemClickListener.onItemClick(fruitEntry) }
        }
    }

    interface OnItemClickListener {
        fun onItemClick(fruitEntry: FruitEntry)
    }
}
